File: fe1-web/model/network/method/message/data/election/EndElection.ts
import {
  Hash, Timestamp,
} from 'model/objects';
import { ProtocolError } from 'model/network/ProtocolError';
import { validateDataObject } from 'model/network/validation';
import { ActionType, MessageData, ObjectType } from '../MessageData';
import { checkTimestampStaleness } from '../Checker';

/** Data sent to end an Election event */
export class EndElection implements MessageData {
  public readonly object: ObjectType = ObjectType.ELECTION;

  public readonly action: ActionType = ActionType.END;

  public readonly lao: Hash;

  public readonly election: Hash;

  public readonly created_at: Timestamp;

  public readonly registered_votes: Hash;

  constructor(msg: Partial<EndElection>) {
    if (!msg.election) {
      throw new ProtocolError('Undefined \'id\' parameter encountered during \'EndElection\'');
    }

    if (!msg.lao) {
      throw new ProtocolError('Undefined \'lao\' parameter encountered during \'EndElection\'');
    }
    this.lao = msg.lao;

    if (!msg.created_at) {
      throw new ProtocolError('Undefined \'created_at\' parameter encountered during \'EndElection\'');
    }
    checkTimestampStaleness(msg.created_at);
    this.created_at = msg.created_at;
    if (!msg.registered_votes) {
      throw new ProtocolError('Undefined \'registered_votes\' parameter encountered during \'EndElection\'');
    }

    this.registered_votes = msg.registered_votes;

    if (!msg.election) {
      throw new ProtocolError('Invalid \'election\' parameter encountered during \'EndElection\'');
    }
    this.election = msg.election;
  }

  /**
   * Creates an EndElection object from a given object
   * @param obj
   */
  public static fromJson(obj: any): EndElection {
    const { errors } = validateDataObject(ObjectType.ELECTION, ActionType.END, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid election end\n\n${errors}`);
    }

    return new EndElection({
      ...obj,
      created_at: new Timestamp(obj.created_at),
      election: new Hash(obj.election),
      lao: new Hash(obj.lao),
    });
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/election/EndElection.scala
package ch.epfl.pop.model.network.method.message.data.election

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, Timestamp}
import spray.json._

final case class EndElection(
    lao: Hash,
    election: Hash,
    created_at: Timestamp,
    registered_votes: Hash
) extends MessageData {
  override val _object: ObjectType = ObjectType.ELECTION
  override val action: ActionType = ActionType.END
}

object EndElection extends Parsable {
  def apply(
      lao: Hash,
      election: Hash,
      created_at: Timestamp,
      registered_votes: Hash
  ): EndElection = new EndElection(lao, election, created_at, registered_votes)

  override def buildFromJson(payload: String): EndElection = payload.parseJson.asJsObject.convertTo[EndElection]
}
